### DISCUSSION TITLE
Airspy Mini + Spyverter (upconverter) should work with dumphfdl?

### DISCUSSION THREAD

**Discussion by dlpwx** [szpajder/dumphfdl#9] [Q&A] [answered]
I'm not having luck getting any messages with an Airspy Mini + Spyverter upconverter using --freq-offset or other attempted work-arounds. Per the readme, is it possible the 3.0 MSPS sample rate is just too large? I've tried various versions of options but have not been able to receive a single message yet. At 3.0 MSPS, CPU core utilization is just below 100%, not a problem on this multi-core Linux desktop PC.

I have excellent dumphfdl performance with the Airspy HF+ Discovery and my wire dipole antenna for a couple weeks now. I was hoping the old Airspy Mini + Spyverter would work okay for a second receiver, but I can't figure out why it won't work at all.  The Airspy Mini + Spyverter works just great in SDR# on Windows. I can see and hear the HFDL signals just as well with AirspyHF+ or Mini + Spyverter, but no luck  getting hfdl messages with Mini + Spyverter on Linux with DumpHFDL yet. I even tried matching linearity gain with what looks best in SDR#. I tried different flavors of freq-offset and stuff. I must be missing something obvious?

I tried powering the Spyverter by USB in case it was a SoapySDR bias tee issue, but that didn't help. So I ponder, is it as simple as DumpHFDL needing a narrower sample rate like 256000 which I use with the AirspyHF+?

It starts fine, connects to VRS, no errors. Just no messages.

`  --soapysdr driver=airspy,biastee=true,serial=<serial_num> --gain-elements LNA=0,MIX=2,VGA=8 --sample-rate 3000000 5502 5508 5514 5451 5529 5538 5544 5547 5583 5589 5622 5652 5655 --freq-offset 120000`

```
dumpvdl2 1.1.0-4e789df (libacars 2.1.3)
driver=airspy,biastee=true,serial=serial_num: computed center frequency: 5553.000 kHz
System table loaded from /home/me/hfdl/systable.conf
Found device #0:
  driver = airspy
  label = AirSpy One [<serial_num>]
  serial = <serial_num>
driver=airspy,biastee=true,serial=<serial_num>: center frequency set to 125553.000 kHz
driver=airspy,biastee=true,serial=<serial_num>: frequency correction set to 0.00 ppm
driver=airspy,biastee=true,serial=<serial_num>: gain element LNA set to 0.00 dB
driver=airspy,biastee=true,serial=<serial_num>: gain element MIX set to 2.00 dB
driver=airspy,biastee=true,serial=<serial_num>: gain element VGA set to 8.00 dB
driver=airspy,biastee=true,serial=<serial_num>: using antenna RX
driver=airspy,biastee=true,serial=<serial_num>: using native sample format CS16 (full_scale: 32767.000)
[INFO] Using format CS16.
/home/me/hfdl/basestation.sqb: database opened
Connecting to xyz etc
```

```
$ dumphfdl --version
dumpvdl2 1.1.0-4e789df (libacars 2.1.3)
```

```
Probe device

----------------------------------------------------
-- Device identification
----------------------------------------------------
  driver=Airspy
  hardware=Airspy
  serial=<serial_num>

----------------------------------------------------
-- Peripheral summary
----------------------------------------------------
  Channels: 1 Rx, 0 Tx
  Timestamps: NO
  Other Settings:
     * Bias tee - Enable the 4.5v DC Bias tee to power SpyVerter / LNA / etc. via antenna connection.
       [key=biastee, default=false, type=bool]
     * Bit Pack - Enable packing 4 12-bit samples into 3 16-bit words for 25% less USB trafic.
       [key=bitpack, default=false, type=bool]

----------------------------------------------------
-- RX Channel 0
----------------------------------------------------
  Full-duplex: NO
  Supports AGC: YES
  Stream formats: CS16, CF32
  Native format: CS16 [full-scale=32767]
  Antennas: RX
  Full gain range: [0, 45] dB
    LNA gain range: [0, 15] dB
    MIX gain range: [0, 15] dB
    VGA gain range: [0, 15] dB
  Full freq range: [24, 1800] MHz
    RF freq range: [24, 1800] MHz
  Sample rates: 6, 3 MSps

```

**Comment by szpajder** [comment]
I believe it should work, however I don't have a spyverter at hand to verify this in practice.

From the output that you've posted it seems that it has set the correct center frequency (5553.000 kHz computed + 120 MHz offset = 125553.000 kHz). Now the question is whether the tuning is precise enough. Maybe there is a residual offset that is too large for dumphfdl to compensate. An offset of 1 ppm at HF is just a couple of Hertz of error - no big deal. However 1 ppm at 125 MHz is 125 Hz of error, which is too much. If this is the case, `--freq-correction` option might help.

Please grab [rx_tools](https://github.com/rxseger/rx_tools) and take an I/Q recording with the exact same setup as you use with dumphfdl. That would be probably something like:

```
rx_sdr -d driver=airspy,biastee=true,serial=<serial_num> -s 3000000 -f 125553000 -F CF32 hfdl.cf32
```

Don't go crazy with the length, as the file will get large pretty quickly. Just make sure that there are any HFDL frames in it (eg. monitor the band with another receiver or some nearby websdr, etc). You can then load the file into [inspectrum](https://github.com/miek/inspectrum) and verify if it makes sense. Or just put it up somewhere for download, so that I could get a glimpse on it.

**Comment by dlpwx** [comment]
Thanks for the helpful advice. I was able to capture samples. The samples do look different between Airspy Mini with Spyverter and AirspyHF+ Discovery in terms of power/noise, especially the surrounding frequencies. Min/Max Power sliders in Inspectrum more to the right with Airspy Mini/Spyverter vs AirspyHF+. The frequency lines 300 to 1000 kHz away from center are much more pronounced with Airspy Mini/Spyverter. Pretty much invisible in AirspyHF+ samples when viewed in Inspectrum.

 Yet, the Airspy Mini + Spyverter signal is visible and mostly accurate / similar pattern as AirspyHF+ samples. 

I tried all sorts of things, very low gain, PPM corrections, and manual frequency adjustments. I have yet to log a single Dumphfdl message with the Airspy Mini & Spyverter. Outside of dumphfdl, the HFDL signal sounds clear and seen frequently. Though I imagine apps like SDR# do some filtering in addition to what I can turn off in the UI. 

Just in case, here's a really short sample with an HFDL signal via Airspy Mini w/Spyverter. (rx_sdr -d driver=airspy,biastee=true -s 3000000 -g LNA=1,MIX=0,VGA=6 -f 131387000 -I CS16 -F CF32 hfdl8.cf32)
50 MB zipped sample --> https://ufile.io/hda9xq72

Performance of AirspyHF+ Discovery continues to great with Dumphfdl though!

**Comment by szpajder** [comment]
This sample has a tuning error of approx. +450 Hz. You can decode it with:

```
dumphfdl --iq-file sample5.cf32 --sample-rate 3000000 --sample-format cf32 --centerfreq 0.45 0
```

Apply `--freq-correction 3.6` or thereabouts and you should be good. Once you get something decoded, you may tune the value further by trial and error to get the mean frequency error down to zero (the offset is displayed in the header of each frame).

**Reply by dlpwx** [reply]
Thanks again. I'm able to decode samples as well, but, even with PPM or freq-correction adjustments, still no luck decoding a live stream with the native output from Airspy Mini. 

--Edited my reply here as I was able to record cs16 samples with rx_sdr and decode them with dumphfdl. 

Recording and decoding samples works.
e.g: ` rx_sdr -d driver=airspy,biastee=true -g LNA=3,MIX=0,VGA=10 -s 3000000 -f 126559000 -I CS16 -F CS16 hfdl.cs16`
decodes with: `dumphfdl --iq-file hfdl.cs16 --sample-rate 3000000 --sample-format cs16 --centerfreq 0.45 0` 

But I can't get below to decode/log any messages. Tried PPM 3.55 to 3.6 and manually setting target freq + .450 kHz. Also tried removing the -output options and using a single frequency. I'm still stumped on this one!

```
dumphfdl --output decoded:basestation:tcp:address=192.168.7.110,port=30014 --output decoded:text:file:path=/home/me/hfdl/hfdl3.log,rotate=daily --prettify-xml --output-mpdus --bs-db /home/me/hfdl/basestation.sqb --ac-details normal --system-table /home/me/hfdl/systable3.conf --soapysdr driver=airspy,biastee=true --gain-elements LNA=1,MIX=0,VGA=6 --sample-rate 3000000 11184 11306 11312 11318 11321 11327 11348 11354 11384 11387 13264 13270 13276 13303 13312 13315 13321 13342 13351 13354 --freq-offset 120000 --freq-correction 3.6

dumpvdl2 1.1.0-4e789df (libacars 2.1.3)
driver=airspy,biastee=true: computed center frequency: 12269.000 kHz
System table loaded from /home/me/hfdl/systable3.conf
Found device #0:
  driver = airspy
  label = AirSpy One [serial]
  serial = serial
driver=airspy,biastee=true: center frequency set to 132269.000 kHz
driver=airspy,biastee=true: frequency correction set to 3.60 ppm
driver=airspy,biastee=true: gain element LNA set to 1.00 dB
driver=airspy,biastee=true: gain element MIX set to 0.00 dB
driver=airspy,biastee=true: gain element VGA set to 6.00 dB
driver=airspy,biastee=true: using antenna RX
driver=airspy,biastee=true: using native sample format CS16 (full_scale: 32767.000)
[INFO] Using format CS16.
/home/me/hfdl/basestation.sqb: database opened
Connecting to 192.168.7.110:30014...
Connection to 192.168.7.110:30014 established

```

```
 SoapySDRUtil --probe
######################################################
##     Soapy SDR -- the SDR abstraction library     ##
######################################################

Probe device

----------------------------------------------------
-- Device identification
----------------------------------------------------
  driver=Airspy
  hardware=Airspy
  serial=serial

----------------------------------------------------
-- Peripheral summary
----------------------------------------------------
  Channels: 1 Rx, 0 Tx
  Timestamps: NO
  Other Settings:
     * Bias tee - Enable the 4.5v DC Bias tee to power SpyVerter / LNA / etc. via antenna connection.
       [key=biastee, default=false, type=bool]
     * Bit Pack - Enable packing 4 12-bit samples into 3 16-bit words for 25% less USB trafic.
       [key=bitpack, default=false, type=bool]

----------------------------------------------------
-- RX Channel 0
----------------------------------------------------
  Full-duplex: NO
  Supports AGC: YES
  Stream formats: CS16, CF32
  Native format: CS16 [full-scale=32767]
  Antennas: RX
  Full gain range: [0, 45] dB
    LNA gain range: [0, 15] dB
    MIX gain range: [0, 15] dB
    VGA gain range: [0, 15] dB
  Full freq range: [24, 1800] MHz
    RF freq range: [24, 1800] MHz
  Sample rates: 6, 3 MSps
```

**Comment by szpajder** [comment, marked as answer]
I checked with my Airspy Mini and it seems that this device does not support applying frequency correction. For comparison, here is what `SoapySDR --probe` shows for an RTL-SDR dongle:

```
    Full freq range: [23.999, 1764] MHz
      RF freq range: [24, 1764] MHz
      CORR freq range: [-0.001, 0.001] MHz
```

Applying a frequency correction is done by setting the CORR frequency component to the given value, however an Airspy device does not provide one. So what you can do instead is cheat on your upconversion frequency. Given that your setup tunes about 450 Hz too high on 125 MHz, you could try setting `--freq-offset` to 119999.55. That should set the tuner slightly lower and compensate the error. The drawback is that the offset value will vary with the tuned frequency. 

Of course you may omit the `--freq-correction` option altogether, since it does nothing for this device.

As far as Airspy calibration is concerned, the web page says that R2 units are factory calibrated down to ~0.05 ppm of error. Dunno if this statement applies to Mini units too, however I checked mine at VHF and it is spot on. So I guess the offset you are seeing comes from the Spyverter. There is an R2x version with an external clock input (https://v3.airspy.us/product/spyverter-r2x/) - probably introduced for a reason.

**Reply by dlpwx** [reply]
--freq-offset to 119999.55 works! (with small tweaks for used frequency). I tried this yesterday but perhaps my math or another setting was off. 

This works fine for me as a second HFDL receiver. With the large sample rate, it can stay set to listen to a large number of often active HFDL frequencies. 

Thanks for all the quick help and another really enjoyable project.

**Reply by dlpwx** [reply]
PS. Even at 6.0 MSPS messages decode really well across many freqs on a desktop Linux machine. Messages decode every couple seconds. I might keep it there.

**Reply by szpajder** [reply]
In fact the program has already been tested with rates up to 10 Msps. There are no hardcoded limits - it's just about how much horsepower your hardware has and whether the abundance of strong signals in the monitored band doesn't saturate the radio.

See https://acars-vdl2.groups.io/g/HFDL/message/102 for some more details on the experiment.

**Comment by LittleDummy420** [comment]
I tried getting my airspy mini and spyverter to work, and tried various offsets with no luck and gave up q.q

  (using the adsb.im image)
I was able to see messages in sdr++ some faint some strong, but was unable to decode any :(

@dlpwx
Did you continue to have success with this? or make any other tweaks?

**Reply by LittleDummy420** [reply]
<img width="1277" height="1300" alt="image" src="https://github.com/user-attachments/assets/75716b7b-c7f1-48ab-8c2b-ad80d3d2805c" />
